stdlib/public/core/HashedCollections.swift.gyb

```diff
@@ -268,20 +268,15 @@ internal var _hashContainerDefaultMaxLoadFactorInverse: Double {
 internal func _stdlib_NSObject_isEqual(_ lhs: AnyObject, _ rhs: AnyObject) -> Bool
 #endif
 
-//===--- Hacks and workarounds --------------------------------------------===//
-
-// FIXME(ABI)#158 (Performance)
-// There used to be a now-fixed optimizer bug that prevented using
-// UnsafePointer<Unmanaged<AnyObject>> in Dictionary implementation.
-// rdar://problem/16861508
-/// Like `UnsafeMutablePointer<Unmanaged<AnyObject>>`, or `id
-/// __unsafe_unretained *` in Objective-C ARC.
+/// A convenience for using `UnsafeMutablePointer<Unmanaged<AnyObject>>`
+/// (`id __unsafe_unretained *` in Objective-C ARC).
 internal struct _UnmanagedAnyObjectArray {
   /// Underlying pointer.
-  internal var value: UnsafeMutableRawPointer
+  internal var value: UnsafeMutablePointer<Unmanaged<AnyObject>>
 
   internal init(_ up: UnsafeMutablePointer<AnyObject>) {
-    self.value = UnsafeMutableRawPointer(up)
+    self.value = UnsafeMutableRawPointer(up).assumingMemoryBound(
+        to: Unmanaged<AnyObject>.self)
```

Comment:
The optimizer is not taking advantage of the capacity. The user is making a contract with the language. The memory's bound type is a property of the memory not of the pointer value.
```
ptrA: UnsafePointer<A>
let ptrB = UnsafeRawPointer(ptrA).bindMemory(B.self)
let a = ptrA[1] // Is this undefined behavior? (yes if capacity > 1)
let b = ptrB[1] // Is this undefined behavior? (yes if capacity < 2)
```